File: python/existing_calculation.py
from rpla import RPLA
from literal import Literal

class ExistingCalculation:
    def __init__(self, rpla, cost_calc=None):
        self.rpla = rpla
        self.functions = rpla.functions
        self.products = rpla.products
        self.finalProducts = []
        self.literals = []
        self.quantumCost = 0
        self.garbages = 0
        self.gates = 0
        self.delay = 0
        self.cost_calc = cost_calc

    def funProRearrange(self):
        mapping = list(range(len(self.products)))
        self.products.sort(key=lambda p: p.getFrequency())
        for i, product in enumerate(self.products):
            mapping[product.id] = i
        for function in self.functions:
            function.productsList = [mapping[pos] for pos in function.productsList]
        for i in range(len(self.functions) - 1, 0, -1):
            if self.functions[i - 1].getSize() == 1:
                self.functions[i], self.functions[i - 1] = self.functions[i - 1], self.functions[i]
                break

    def xorPlane(self):
        xorTDOT = 0
        totalXOROperation = 0
        self.funProRearrange()
        self.resetFlagOfProduct()
        for function in reversed(self.functions):
            totalXOROperation += function.getSize() - 1
            if function.getSize() == 1:
                if not self.products[function.productsList[0]].flag:
                    xorTDOT += 1
                    self.products[function.productsList[0]].flag = True
            else:
                flag = 0
                for prod_idx in function.productsList:
                    if not self.products[prod_idx].flag:
                        if flag == 0:
                            xorTDOT += 1
                            flag = 1
                        self.products[prod_idx].flag = True
        self.resetFlagOfProduct()
        count = 0
        for function in reversed(self.functions):
            function.productsList.sort(reverse=True)
            if function.productsList:
                count = self.products[function.productsList[0]].delayCountXOR
            else:
                count = 0
            flag = 0
            for j, prod_idx in enumerate(function.productsList):
                if not self.products[prod_idx].flag:
                    flag = 1
                    self.products[prod_idx].flag = True
                if self.products[prod_idx].delayCountXOR + 1 > count:
                    count = self.products[prod_idx].delayCountXOR + 1
                else:
                    count += 1
                if j == function.getSize() - 1 and flag == 1:
                    count -= 1
                self.products[prod_idx].delayCountXOR = count
        self.gates = totalXOROperation + len(self.functions) - xorTDOT
        self.garbages = len(self.products) - xorTDOT
        self.quantumCost = self.gates
        print("==========================================================")
        print("                Existing Calculation")
        print("==========================================================")
        print("      Rearranged FUNCTIONS & PRODUCTS ")
        print("==========================================================")
        self.showFunctions()
        self.showProducts()
        print("==========================================================")
        print("                     EXOR Plane")
        print("==========================================================")
        print(f"Total EXOR Operations : {totalXOROperation}")
        print(f"TDOT                  : {xorTDOT}")
        print(f"Feynman Gate          : {self.gates}")
        print(f"Garbage, GB           : {self.garbages}")
        print("==========================================================")

    def andPlane(self):
        andTDOT = 0
        totalANDOperation = 0
        constantProductExist = False
        self.initializeDoubleLiteral()
        for i in range(len(self.products) - 1, -1, -1):
            tempProduct = self.products[i]
            flag = False
            if tempProduct.getSize() > 0:
                for index in tempProduct.literals:
                    if not flag:
                        andTDOT += 1
                        self.literals[index].dotFlag = True
                    flag = True
                    self.literals[index].flag = True
                totalANDOperation += tempProduct.getSize() - 1
            else:
                constantProductExist = True
        for tempLiteral in self.literals:
            count = 0
            if tempLiteral.productsList:
                count = self.products[tempLiteral.productsList[0]].delayCountAND
            for j, prod_idx in enumerate(tempLiteral.productsList):
                if self.products[prod_idx].delayCountAND + 1 > count:
                    count = self.products[prod_idx].delayCountAND + 1
                else:
                    count += 1
                if j == len(tempLiteral.productsList) - 1 and tempLiteral.dotFlag:
                    count -= 1
                self.products[prod_idx].delayCountAND = count
        if constantProductExist:
            totalXOROperation = len(self.products) - andTDOT - 1
        else:
            totalXOROperation = len(self.products) - andTDOT
        totalXOROperation += len(self.literals) // 2
        totalGarbages = totalANDOperation + 2 * (len(self.literals) // 2) - andTDOT
        self.quantumCost += totalANDOperation * 5 + totalXOROperation
        self.delay = 0
        for product in self.products:
            if product.getSize() != 0:
                product.delayCountAND += 1
            self.delay = max(self.delay, product.delayCountAND + product.delayCountXOR)
        self.gates += totalANDOperation + totalXOROperation
        self.garbages += totalGarbages
        print("==========================================================")
        print("                    AND Plane")
        print("==========================================================")
        print(f"Total AND Operations: {totalANDOperation}")
        print(f"TDOT                : {andTDOT}")
        print(f"Total Toffoli Gates : {totalANDOperation}")
        print(f"Total Feynman Gates : {totalXOROperation}")
        print(f"Garbage, GB         : {totalGarbages}\n")
        print("==========================================================")
        print("                  Delay ")
        print("==========================================================")

    def showFinalResult(self):
        print("==========================================================")
        print(f"           Final Calculation of {self.rpla.esopFileName}")
        print("==========================================================")
        if self.cost_calc is not None:
            print(f"Total Gates   : {self.gates}({self.cost_calc.gates})")
            print(f"Total Garbages: {self.garbages}({self.cost_calc.garbages})")
            print(f"Total Delay   : {self.delay}({self.cost_calc.delay})")
            print(f"Total Q. Cost : {self.quantumCost}({self.cost_calc.quantumCost})")
        else:
            print(f"Total Gates   : {self.gates}")
            print(f"Total Garbages: {self.garbages}")
            print(f"Total Delay   : {self.delay}")
            print(f"Total Q. Cost : {self.quantumCost}")

    def initializeDoubleLiteral(self):
        self.literals = []
        for i in range(len(self.products)):
            self.products[i].literals.clear()
            item = self.products[i].bitPattern
            for j in range(2 * RPLA.totalLiterals):
                if i == 0:
                    self.literals.append(Literal(j))
                if item[j // 2] == '1' and j % 2 == 0:
                    self.products[i].literals.append(j)
                    self.literals[j].addProduct(i)
                elif item[j // 2] == '0' and j % 2 == 1:
                    self.products[i].literals.append(j)
                    self.literals[j].addProduct(i)

    def resetFlagOfProduct(self):
        for product in self.products:
            product.flag = False
            product.delayCountAND = 0
            product.delayCountXOR = 0

    def showProducts(self):
        print("----------------------------------------------------------")
        for product in self.products:
            line = f"Product [{product.id}]: "
            line += ",".join(f"f{j}({pos})" for j, pos in enumerate(product.functions) if pos >= 0)
            print(line)
        print("----------------------------------------------------------")

    def showFunctions(self):
        print("----------------------------------------------------------")
        for function in self.functions:
            line = f"Function [{function.id}]: "
            line += ",".join(f"P{pos}" for pos in function.productsList)
            print(line)
        print("----------------------------------------------------------")

File: javacode/benchmarkpla/ExistingCalculation.java
/*********************************************************
 *          Author : Sajib Kumar Mitra
 *         Position: M.Sc.(1st Semester)/2008-09
 *       Department: Computer Science and Engineering
 *        Institute: University of Dhaka
 *            Email: sajibmitra.csedu@yahoo.com
 *            Phone: +8801710283239
 *********************************************************/
package benchmarkpla;

import java.util.ArrayList;
import java.util.Collections;
public class ExistingCalculation {

    private static int gates;               //Total number of gates
    private static int garbages;            //Total number of garbages
    private static int delay;               //delay calculation
    private static int quantumCost;         //Quantum Cost Calculation
    public ArrayList<Function> functions;   //store functions
    public ArrayList<Product> products;     //store products
//finalProducts store the ordered series of Products
    public ArrayList<Product> finalProducts;
    public ArrayList<Literal> literals;
    public Function tempFunction;
    public Product  tempProduct;
    public Literal  tempLiteral;
//xorTDOT for XOR plane and andTDOT for AND plane
    private int xorTDOT;
    private int andTDOT;
    private Boolean constantProductExist;
//this constructor initializes the functions and products by using initial status
    public ExistingCalculation(){
        functions=RPLA.functions;
        products=RPLA.products;
        finalProducts=new ArrayList<Product>();
        literals=new ArrayList<Literal>();
        quantumCost=0;
        garbages=0;
        gates=0;
        delay=0;
    }
//XOR Plane calculates the total number of gates and garbages
    public void funProRearrange(){
        ArrayList <Integer> mapping=new ArrayList<Integer>();
        for(int i=0;i<RPLA.totalProducts;i++){
            mapping.add(i);
        }
        for(int i=0;i<RPLA.totalProducts-1;i++)
            for(int j=i+1;j<RPLA.totalProducts;j++){
                if(products.get(i).getFrequency()>products.get(j).getFrequency()){
                     tempProduct=products.get(i);
                     products.set(i, products.get(j));
                     products.set(j, tempProduct);                                          
                }
            }
        for(int i=0;i<RPLA.totalProducts;i++){
            mapping.set(products.get(i).id,i);
        }        
        for(int i=0; i<RPLA.totalOutputs;i++){
                tempFunction=functions.get(i);
                for(int j=0;j<tempFunction.getSize();j++){
                    tempFunction.productsList.set(j,mapping.get(tempFunction.productsList.get(j)));
                }
        }
  //      showFunctions();
        for(int i=RPLA.totalOutputs-1;i>=1;i--)
            for(int j=i-1;j>=0;j--)
                if(functions.get(j).getSize()==1){
                    tempFunction=functions.get(i);
                    functions.set(i, functions.get(j));
                    functions.set(j, tempFunction);
                    break;
                }
        }
    public void xorPlane(){
        xorTDOT=0;
        int totalXOROperation=0;
        funProRearrange();        
        resetFlagOfProduct();
        for(int i=RPLA.totalOutputs-1;i>=0;i--){
            tempFunction=functions.get(i);
             totalXOROperation+=tempFunction.getSize()-1;
            if(tempFunction.getSize()==1){
                if(!products.get(tempFunction.productsList.get(0)).flag){
                    xorTDOT++; products.get(tempFunction.productsList.get(0)).flag=true;
                }
                else { }
            }
            else{
                int flag=0;
                for(int j=0;j<tempFunction.getSize();j++){
                    if(!products.get(tempFunction.productsList.get(j)).flag){
                       if(flag==0){flag=1;xorTDOT++;} else{}
                           products.get(tempFunction.productsList.get(j)).flag=true;
                    }
                    else{ }
                }
            }
        }
        resetFlagOfProduct();
        int count=0;
        for(int i=RPLA.totalOutputs-1;i>=0;i--){
            tempFunction=functions.get(i);
            Collections.sort(tempFunction.productsList,Collections.reverseOrder());
            if(tempFunction.productsList.size()!=0)
                count=products.get(tempFunction.productsList.get(0)).delayCountXOR;
            else
                count=0;
            int flag=0;
            for(int j=0;j<tempFunction.getSize();j++){
                if(!products.get(tempFunction.productsList.get(j)).flag){ //If any product has not used yet
                    flag=1; products.get(tempFunction.productsList.get(j)).flag=true;
                }
                if(products.get(tempFunction.productsList.get(j)).delayCountXOR+1>count)
                    count=products.get(tempFunction.productsList.get(j)).delayCountXOR+1;
                else
                    count++;
                if(j==tempFunction.getSize()-1&&flag==1){
                       count--;
                }
                products.get(tempFunction.productsList.get(j)).delayCountXOR=count;
             }
        }
        gates=totalXOROperation+RPLA.totalOutputs-xorTDOT;
        garbages=RPLA.totalProducts-xorTDOT;
        quantumCost=gates;
//remove the comment symbol If U want to see detail..
        System.out.println("==========================================================");
        System.out.println("                Existing Calculation");
        System.out.println("==========================================================");
        System.out.println("      Rearranged FUNCTIONS & PRODUCTS ");
        System.out.println("==========================================================");
                      showFunctions();
                      showProducts();
        System.out.println("==========================================================");
        System.out.println("                     EXOR Plane");
        System.out.println("==========================================================");
        System.out.println("Total EXOR Operations : "+ totalXOROperation);
        System.out.println("TDOT                  : "+xorTDOT);
        System.out.println("Feynman Gate          : "+gates);
        System.out.println("Garbage, GB           : "+garbages);
        System.out.println("==========================================================");
    }
    public void andPlane(){
    Boolean flag;
        andTDOT=0;
        int totalANDOperation=0;
        int totalXOROperation=0;
        int totalGarbages =0;
        constantProductExist=Boolean.FALSE;
        initializeDoubleLiteral();
        for(int i=RPLA.totalProducts-1;i>=0;i--){
            tempProduct=products.get(i);
            flag=false;
            if(tempProduct.getSize()>0){
                for(int j=0;j<tempProduct.getSize();j++){
                    int index =tempProduct.literals.get(j);
                    if(!flag&&!literals.get(index).flag){
                        andTDOT++;
                        literals.get(index).dotFlag=Boolean.TRUE; // already used and has dot
                    }
                    flag=true;
                    literals.get(index).flag=true;  // already used but doesn't have dot
                }
                totalANDOperation+=tempProduct.getSize()-1;
            }
            else
              constantProductExist=Boolean.TRUE;
        }
        int count=0;
        for(int i=0;i<2*RPLA.totalLiterals;i++){
            tempLiteral=literals.get(i);

            if(tempLiteral.getSize()!=0)
            count=products.get(tempLiteral.productsList.get(0)).delayCountAND;
            else
            count=0;
            for(int j=0;j<tempLiteral.getSize();j++){
                if(products.get(tempLiteral.productsList.get(j)).delayCountAND+1>count)
                    count=products.get(tempLiteral.productsList.get(j)).delayCountAND+1;
                else
                    count++;
                if(j==(tempLiteral.getSize()-1)&& tempLiteral.dotFlag)
                    count--;
                else{}
               products.get(tempLiteral.productsList.get(j)).delayCountAND=count;
             }
        }/*
        if(constantProductExist)
            totalXOROperation=RPLA.totalProducts-andTDOT-1;
        else
            totalXOROperation=RPLA.totalProducts-andTDOT;
        totalXOROperation+=RPLA.totalLiterals;
        totalGarbages=totalANDOperation+2*RPLA.totalLiterals-andTDOT;
        quantumCost+=totalANDOperation*5+totalXOROperation;
        delay=0;
        for(int j=0;j<RPLA.totalProducts;j++){
            if(products.get(j).getSize()!=0)
                products.get(j).delayCountAND++;
                if(delay<(products.get(j).delayCountAND+products.get(j).delayCountXOR)){
                    delay=products.get(j).delayCountAND+products.get(j).delayCountXOR;
                }
            }
        gates+=totalANDOperation+totalXOROperation;
        garbages+=totalGarbages;*/
//Remove the code from "If(constantProductExist)" to here and remove the comments are shown in follows for detail view..

        System.out.println("==========================================================");
        System.out.println("                    AND Plane");
        System.out.println("==========================================================");
        System.out.println("Total AND Operations: " +totalANDOperation);
        System.out.println("TDOT                : "+andTDOT);
        System.out.println("Total Toffoli Gates : "+totalANDOperation);
        if(constantProductExist)
            totalXOROperation=RPLA.totalProducts-andTDOT-1;
        else
            totalXOROperation=RPLA.totalProducts-andTDOT;
        totalXOROperation+=RPLA.totalLiterals;
        System.out.println("Total Feynman Gates : "+totalXOROperation);
            totalGarbages=totalANDOperation+2*RPLA.totalLiterals-andTDOT;
        System.out.println("Garbage, GB         : "+totalGarbages+"\n");
        System.out.println("==========================================================");
        System.out.println("                  Delay ");
        System.out.println("==========================================================");
        System.out.println("    Delay (AND) |  Delay (EXOR)   = Total Delay");
        System.out.println("==========================================================");
        quantumCost+=totalANDOperation*5+totalXOROperation;
        delay=0;
        for(int j=0;j<RPLA.totalProducts;j++){
            if(products.get(j).getSize()!=0)
            products.get(j).delayCountAND++;
                System.out.println("P"+ products.get(j).id+" :     "+products.get(j).delayCountAND+"              "+products.get(j).delayCountXOR +"            "+ (products.get(j).delayCountAND+products.get(j).delayCountXOR));
                if(delay<(products.get(j).delayCountAND+products.get(j).delayCountXOR)){
                    delay=products.get(j).delayCountAND+products.get(j).delayCountXOR;
                }
            }
        gates+=totalANDOperation+totalXOROperation;
        garbages+=totalGarbages;
    }
    public void showFinalResult(){
      System.out.println("==========================================================");
      System.out.println("           Final Calculation of "+ RPLA.esopFileName);
      System.out.println("==========================================================");
      System.out.println("Total Gates   : "+gates +"("+CostCalculation.gates+")");
      System.out.println("Total Garbages: "+garbages+"("+CostCalculation.garbages+")");
      System.out.println("Total Delay   : "+delay+"("+CostCalculation.delay+")");
      System.out.println("Total Q. Cost : "+quantumCost+"("+CostCalculation.quantumCost+")");
      System.out.println("==========================================================");
    }
    public void initializeDoubleLiteral(){
        for(int i=0; i<RPLA.totalProducts;i++){
            products.get(i).literals.clear();
            String item=products.get(i).bitPattern;
         for(int j=0;j<2*RPLA.totalLiterals;j++){
            if(i==0) literals.add(new Literal(j));
            else {}
            if(item.charAt(j/2)=='1'&&(j%2==0)){
              products.get(i).literals.add(j);
              literals.get(j).addProduct(i);
            }
            else if(item.charAt(j/2)=='0'&&(j%2==1)){
                products.get(i).literals.add(j);
                literals.get(j).addProduct(i);
            }
            else{}
         }
       }
    }
    
    public void resetFlagOfProduct(){
       for(int j=0;j<RPLA.totalProducts;j++){
                products.get(j).flag=false;
                products.get(j).delayCountAND=0;
                products.get(j).delayCountXOR=0;
            }
    }
    public int getUpdatedIdOfPro(int previousId){
        int i;
        for(i=0;i<finalProducts.size();i++){
            if(finalProducts.get(i).id==previousId){
             break;
            }
        }
        return i;
    }
    
    public void sortProductListOfFunc(){
        for(int i=0;i<RPLA.totalOutputs;i++){
                 Collections.sort(functions.get(i).productsList);
        }
    }
    public void showProducts(){
        System.out.println("----------------------------------------------------------");
            for(int i=0;i<RPLA.totalProducts;i++){
                         System.out.print ("Product ["+products.get(i).id+"]: ");
                         for(int j=0;j<products.get(i).functions.size();j++){
                                int pos=products.get(i).functions.get(j);
                             if(pos>=0){
                                System.out.print("f"+j+"("+pos+ ") ");
                             }
                         }
                         System.out.println();
                      }
       System.out.println("----------------------------------------------------------");
    }
    public void showFunctions(){
        System.out.println("----------------------------------------------------------");
        for(int i=0;i<RPLA.totalOutputs;i++){
                          System.out.print("Function ["+functions.get(i).id+"]: ");
                          for(int j=0;j<functions.get(i).productsList.size();j++){
                             System.out.print("P"+products.get(functions.get(i).productsList.get(j)).id);
                             if(j!=functions.get(i).productsList.size()-1)
                             System.out.print(",");
                          }
                          System.out.println();
                      }
        System.out.println("----------------------------------------------------------");
    }
}
